### ISSUE TITLE
Fix S2583 FP: casts to nullable values are treated as never null.

### ISSUE THREAD

**Issue by fiotti** [SonarSource/sonar-dotnet#7974]
### Description

False positive [S2583](https://sonarsource.github.io/rspec/#/rspec/S2583/csharp).

### Repro steps

The condition in the `if` never evaluates to `false` in this example.

```csharp
static class Test
{
    class Example
    {
        public static explicit operator string?(Example example) => null;
    }

    public static void Hello() {
        Example example = new Example();
        string? test = (string?)example;

        // S2583: Change this condition so that it doesn't always evaluate to 'false'; some subsequent code is never executed.
        if (test is null)
            Console.WriteLine("test is null");
    }
}
```

### Expected behavior

I didn't expect any diagnostics to be emitted.

### Actual behavior

False positive.

### Known workarounds

```csharp
#pragma warning disable S2583
```

### Related information

* SonarLint for Visual Studio 2022, version 7.2.0.76209
* Visual Studio Professional 2022 (64 bit), version 17.6.3
* dotnet 7.0.304
* MSBuild version 17.6.3+07e294721 for .NET Framework, version 17.6.3.22601
* Windows 11 Pro 22H2 22621.2215

**Comment by mary-georgiou-sonarsource** [comment]
Hello @fiotti,

This issue has been fixed since [sonar-dotnet 9.8](https://github.com/SonarSource/sonar-dotnet/releases/tag/9.8.0.76515).
The version of SonarLint that you are currently using (7.2), comes with version sonar-dotnet 9.7.
In their next release, this issue will be fixed.

Thanks a lot for the report!
